### PR TITLE
[DGFiP] Update my_arr function

### PR DESCRIPTION
**Pull request by ddeclerck**
The `my_arr` function has been updated in "recent" versions of the legacy compiler.

### CODE REVIEW

File: examples/dgfip_c/const.h [hunk 1]
```diff
@@ -95,7 +95,8 @@ extern void free_erreur();
 #ifdef FLG_OPTIM_MIN_MAX
 
 #define my_floor(a)	(floor_g((a) + 0.000001))
-#define my_arr(a)	(floor_g((a) + 0.50005))
+//#define my_arr(a)	(floor_g((a) + 0.50005)) /* Ancienne version (2021) */
+#define my_arr(a)	(((a) < 0.0) ? ceil_g((a) - .50005) : floor_g((a) + .50005))
```
Comment (denismerigoux) [inline examples/dgfip_c/const.h:99]: This needs to be update here as well:
https://github.com/MLanguage/mlang/blob/9c95761d7b7795fdc8d2d6da0f30176027717bf9/src/mlang/backend_ir/bir_interpreter.ml#L109-L115

**Review by denismerigoux — CHANGES_REQUESTED** [review]
Thanks! But we need to keep the official interpret in sync too.

**Comment by ddeclerck** [comment]
@denismerigoux, that should do it.
Passed all tests with all 3 rounding operation modes.
Maybe a bit over-engineered ;)

File: src/mlang/backend_ir/bir_interpreter.mli [hunk 2]
```diff
@@ -118,19 +120,45 @@ module type S = sig
...
+  val evaluate_expr : ctx -> Mir.program -> Bir.expression Pos.marked -> value
+
+  val evaluate_program : Bir.program -> ctx -> int -> ctx
 end
 
-module RegularFloatInterpreter : S
+module FloatDefInterp : S
+
+module FloatMultInterp : S
+
```
Comment (denismerigoux) [inline src/mlang/backend_ir/bir_interpreter.mli:136]: An interpretation guide would be handy here to know how to decipher the names of those interpreters :)

File: src/mlang/backend_ir/bir_interpreter.mli [hunk 3]
```diff
@@ -143,16 +171,29 @@ type value_sort =
   | Interval
   | Rational
 
+(** Rounding operations to use in the interpreter *)
+type round_ops =
+  | RODefault
+  | ROMulti
```
Comment (denismerigoux) [inline src/mlang/backend_ir/bir_interpreter.mli:177]: Some context about where those different rounding operations come from ?

File: src/mlang/backend_ir/bir_roundops.ml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  RoundOpsInterface with type t = N.t = struct
+  type t = N.t
+
+  let truncatef (x : N.t) : N.t = N.floor N.(x +. N.of_float 0.000001)
+
+  let roundf (x : N.t) =
+    let n_0_5 = N.of_float 0.5 in
+    let n_100000_0 = N.of_float 100000.0 in
+    let v1 = N.floor x in
+    let v2 = N.(N.floor (((x -. v1) *. n_100000_0) +. n_0_5) /. n_100000_0) in
```
Comment (denismerigoux) [inline src/mlang/backend_ir/bir_roundops.ml:53]: This is just a freak show, even more embarrassing when it's laid out in beautiful OCaml

File: src/mlang/optimizing_ir/partial_evaluation.ml [hunk 5]
```diff
@@ -312,7 +311,7 @@ let rec partially_evaluate_expr (ctx : partial_ev_ctx) (p : Mir.program)
               from_literal
                 (Bir_interpreter.evaluate_expr p
                    (Pos.same_pos_as (Mir.Binop (op, new_e1, new_e2)) e1)
-                   RegularFloat)
+                   RegularFloat RODefault)
```
Comment (denismerigoux) [inline src/mlang/optimizing_ir/partial_evaluation.ml:314]: Couldn't you get the right rounding mode here by adding a `Cli.rounding_mode = ref ...` global that is set by `driver.ml` ? Sounds like an easy fix.

File: src/mlang/test_framework/test_interpreter.ml [hunk 6]
```diff
@@ -268,11 +269,14 @@ let check_all_tests (p : Bir.program) (test_dir : string) (optimize : bool)
             (Pos.unmark err.Mir.Error.name);
           (successes, failures, code_coverage_acc)
     in
+    let module Interp = (val Bir_interpreter.get_interp value_sort round_ops
```
Comment (denismerigoux) [inline src/mlang/test_framework/test_interpreter.ml:272]: <3 first-class modules for the win

File: src/mlang/test_framework/test_interpreter.ml [hunk 7]
```diff
@@ -281,119 +285,27 @@ let check_all_tests (p : Bir.program) (test_dir : string) (optimize : bool)
...
-              (err, expr, bindings) -> (
-              ( expr,
-                err,
-                match bindings with
-                | [ (v, Bir_interpreter.BigIntInterpreter.SimpleVar l1) ] ->
-                    Some
-                      (v, Bir_interpreter.BigIntInterpreter.value_to_literal l1)
-                | _ -> None ))
-          | _ -> assert false
-          (* should not happen *)
```
Comment (denismerigoux) [inline src/mlang/test_framework/test_interpreter.ml:334]: I love deleted code

File: src/mlang/test_framework/test_interpreter.ml [hunk 8]
```diff
@@ -281,119 +285,27 @@ let check_all_tests (p : Bir.program) (test_dir : string) (optimize : bool)
...
                 | _ -> None ))
           | _ -> assert false
           (* should not happen *)
         in
         report_violated_condition_error bindings expr err
-    | Bir_interpreter.IntervalInterpreter.RuntimeError
-        (Bir_interpreter.IntervalInterpreter.StructuredError (msg, pos, kont), _)
-    | Bir_interpreter.BigIntInterpreter.RuntimeError
-        (Bir_interpreter.BigIntInterpreter.StructuredError (msg, pos, kont), _)
-    | Bir_interpreter.MPFRInterpreter.RuntimeError
```
Comment (denismerigoux) [inline src/mlang/test_framework/test_interpreter.ml:379]: I remember my pain when I wrote those, thanks so much for getting rid of that

File: src/mlang/utils/cli.ml [hunk 9]
```diff
@@ -144,6 +144,20 @@ let precision =
...
 
+let roundops =
+  Arg.(
+    value
+    & opt (some string) (Some "default")
+    & info [ "roundops" ] ~docv:"ROUNDOPS"
+        ~doc:
+          "Rounding operations to use in the interpreter: default, multi, \
+           mainframe<n> (where n is the size in bits of the long type to \
+           simulate). Each corresponds to the behavior of the legacy DGFiP \
```
Comment (denismerigoux) [inline src/mlang/utils/cli.ml:155]: Even if there is doc here doesn't mind being reminded in `bir_roundops.mli`

**Review by denismerigoux — APPROVED** [review]
Perfect, I love it :)

File: src/mlang/optimizing_ir/partial_evaluation.ml (hunk 5 above)
Comment (ddeclerck) [reply src/mlang/optimizing_ir/partial_evaluation.ml:314]: Yeah, I could to that.
How about doing that also for the number representation too ?

Comment (denismerigoux) [reply src/mlang/optimizing_ir/partial_evaluation.ml:314]: Yep you can do that. Those things are not meant to change during compiler execution.

File: src/mlang/backend_ir/bir_roundops.ml (hunk 4 above)
Comment (ddeclerck) [reply src/mlang/backend_ir/bir_roundops.ml:53]: Yeah, sadly we have to mimic this...

File: src/mlang/optimizing_ir/partial_evaluation.ml (hunk 5 above)
Comment (ddeclerck) [reply src/mlang/optimizing_ir/partial_evaluation.ml:314]: Actually, that leads to a cycle between Cli and Bir_interpreter. That requires me to either :
- move the types `round_ops` and  `value_sort` out of Bir_interpreter
- move the global options out of Cli
- something else ?
What solution du you prefer ?

Comment (denismerigoux) [reply src/mlang/optimizing_ir/partial_evaluation.ml:314]: I would be in favor of moving `round_ops` and `value_sorts` into `Cli` to break the cycle.

**Comment by ddeclerck** [comment]
@denismerigoux Can we merge this now ?

**Pull request merged**